src/test/java/net/snowflake/client/jdbc/PreparedStatementIT.java

```diff
@@ -0,0 +1,1721 @@
+        connection.createStatement().execute("ALTER SESSION SET CLIENT_TIMESTAMP_TYPE_MAPPING = " + tsType);
+
+        connection.createStatement().execute("CREATE OR REPLACE TABLE test_prepst_ts (id INTEGER, tz TIMESTAMP)");
+        prepStatement = connection.prepareStatement("INSERT INTO test_prepst_ts(id, tz) VALUES(?,?)");
+
+        // First, run with non-stage binding
+        connection.createStatement().executeQuery("ALTER SESSION SET CLIENT_STAGE_ARRAY_BINDING_THRESHOLD = 0");
+        for (int i = 0; i < timestamps.length; i++)
+        {
+          prepStatement.setInt(1, i);
+          prepStatement.setTimestamp(2, timestamps[i]);
+          prepStatement.addBatch();
+        }
+        countResult = prepStatement.executeBatch();
+        for (int res : countResult)
+        {
+          assertEquals(1, res);
+        }
+
+        Timestamp[] nonStageResult = new Timestamp[timestamps.length];
+        ResultSet rsNonStage =
+            connection.createStatement().executeQuery("SELECT * FROM test_prepst_ts ORDER BY id ASC");
+        for (int i = 0; i < nonStageResult.length; i++)
+        {
+          rsNonStage.next();
+          nonStageResult[i] = rsNonStage.getTimestamp(2);
+        }
+
+        connection.createStatement().execute("DELETE FROM test_prepst_ts WHERE 1=1");
+
+        // Now, run with stage binding
+        connection.createStatement().execute("ALTER SESSION SET CLIENT_STAGE_ARRAY_BINDING_THRESHOLD = 1"); // enable stage bind
+        for (int i = 0; i < timestamps.length; i++)
+        {
+          prepStatement.setInt(1, i);
+          prepStatement.setTimestamp(2, timestamps[i]);
+          prepStatement.addBatch();
+        }
+        countResult = prepStatement.executeBatch();
+        for (int res : countResult)
+        {
+          assertEquals(1, res);
+        }
+
+        Timestamp[] stageResult = new Timestamp[timestamps.length];
+        ResultSet rsStage = connection.createStatement().executeQuery("SELECT * FROM test_prepst_ts ORDER BY id ASC");
+        for (int i = 0; i < stageResult.length; i++)
+        {
+          rsStage.next();
+          stageResult[i] = rsStage.getTimestamp(2);
+        }
+
+        for (int i = 0; i < timestamps.length; i++)
+        {
+          assertEquals("Stage binding timestamp should match non-stage binding timestamp (" + tsType + ")",
+                       nonStageResult[i], stageResult[i]);
+        }
+      }
+
+    }
+    finally
+    {
+      connection.createStatement().execute("DROP TABLE IF EXISTS test_prepst_ts");
+    }
+  }
+
+  @Test
+  @ConditionalIgnore(condition = RunningOnTravisCI.class)
+  public void testStageBatchTimes() throws SQLException
+  {
+    connection = getConnection();
+    Time tMidnight = new Time(0);
+    Time tNeg = new Time(-1);
+    Time tPos = new Time(1);
+    Time tNow = new Time(System.currentTimeMillis());
+    Time tNoon = new Time(12 * 60 * 60 * 1000);
+    Time[] times = new Time[]{
```

Comment:
Same here. Be creative to add more tests.

Reply:
Separate commit